images/virtualization-artifact/pkg/controller/service/ip_address_service.go

```diff
@@ -28,98 +29,95 @@ import (
 )
 
 type IpAddressService struct {
-	logger      logr.Logger
-	ParsedCIDRs []*net.IPNet
+	parsedCIDRs []netip.Prefix
 }
 
 func NewIpAddressService(
 	logger logr.Logger,
 	virtualMachineCIDRs []string,
 ) *IpAddressService {
-	parsedCIDRs := make([]*net.IPNet, len(virtualMachineCIDRs))
+	parsedCIDRs := make([]netip.Prefix, len(virtualMachineCIDRs))
 
 	for i, cidr := range virtualMachineCIDRs {
-		_, parsedCIDR, err := net.ParseCIDR(cidr)
-		if err != nil || parsedCIDR == nil {
+		parsedCIDR, err := netip.ParsePrefix(cidr)
+		if err != nil {
 			logger.Error(err, fmt.Sprintf("failed to parse CIDR %s:", cidr), "err", err)
 			return nil
 		}
 		parsedCIDRs[i] = parsedCIDR
 	}
 
 	return &IpAddressService{
-		logger:      logger.WithValues("service", "IpAddressService"),
-		ParsedCIDRs: parsedCIDRs,
+		parsedCIDRs: parsedCIDRs,
 	}
 }
 
-func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) bool {
-	ip := net.ParseIP(address)
+func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) error {
+	ip, err := netip.ParseAddr(address)
+	if err != nil || !ip.IsValid() {
+		return errors.New("invalid IP address format")
+	}
 
 	if _, ok := allocatedIPs[ip.String()]; !ok {
-		for _, cidr := range s.ParsedCIDRs {
+		for _, cidr := range s.parsedCIDRs {
 			if cidr.Contains(ip) {
 				// available
-				return true
+				return nil
 			}
 		}
 		// out of range
-		return false
+		return ErrIPAddressOutOfRange
 	}
+
 	// already exists
-	return false
+	return ErrIPAddressAlreadyExist
```

Comment:
Fail early: change `!ok` to `ok` and return ErrIPAddressAlreadyExist inside if, not after.

Reply:
Fixed it.